File: 9_abstraction.py
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def draw(self):
        pass

    def display_info(self):
        print("This is a shape.")

class Circle(Shape):
    def draw(self):
        print("This is a circle.")

class Rectangle(Shape):
    def draw(self):
        print("This is a rectangle.")

def main():
    circle = Circle()
    rectangle = Rectangle()

    shapes = [circle, rectangle]

    for s1 in shapes:
        s1.draw()

    for s1 in shapes:
        s1.display_info()

if __name__ == "__main__":
    main()

File: 9_abstraction.cpp
#include <bits/stdc++.h>
using namespace std;

class Shape {
public:
    virtual void draw() = 0;
    void print_info() {
        cout << "This is a shape." << endl;
    }
};

class Circle : public Shape {
public:
    void draw() override {
        cout << "This is a circle." << endl;
    }
};

class Rectangle : public Shape {
public:
    void draw() override {
        cout << "This is a rectangle." << endl;
    }
};

int main() {
    Circle circle;
    Rectangle rectangle;

    Shape* s1 = &circle;
    Shape* s2 = &rectangle;

    s1->draw();
    s2->draw();

    s1->print_info();
    s2->print_info();

    return 0;
}
